README.md
# Multi-dimensional-graph-based-search-customization
Customizing web search results using local data and multi-dimensional graphs.

See [Report](https://github.com/kanishk509/Multi-dimensional-graph-based-search-customization/blob/master/sop_report.pdf) for details of the project.


history.txt
https://www.python.org/
https://en.wikipedia.org/wiki/Ultimate_(sport)
https://towardsdatascience.com/machine-learning-an-introduction-23b84d51e6d0
https://machinelearningmastery.com/what-is-deep-learning/
https://en.wikipedia.org/wiki/Tennis
https://machinelearningmastery.com/machine-learning-in-python-step-by-step/

flask_server/time_track_server.py
from flask import Flask, jsonify,request
import time, tldextract, json, requests

app = Flask(__name__)


file = open('data.json', 'r')
website_dict = json.load(file)
url_timestamp = {}
prev_url = ""


def url_domain(url):
	url_list = tldextract.extract(url)
	domain_name = url_list.domain + '.' + url_list.suffix

	if(domain_name == '.'):
		return "EMPTY_DOMAIN_NAME"
	else:
		return domain_name

dont_track = ["EMPTY_DOMAIN_NAME", "newtab."]


@app.route('/send_url', methods=['POST'])
def send_url():
	resp_json = request.get_data()
	params = resp_json.decode()
	url = params.replace("url_for_time=", "")
	print("Open website: " + url_domain(url))
	parent_url = url_domain(url)

	global url_timestamp
	global website_dict
	global prev_url

	if(parent_url not in website_dict.keys()) and (parent_url not in dont_track):
		website_dict[parent_url] = {'time_spent' : 0, 'score' : 0, 'topic' : ""}

	if(prev_url != '') and (prev_url not in dont_track):
		time_spent = int(time.time() - url_timestamp[prev_url])
		website_dict[prev_url]['time_spent'] += time_spent

	x = int(time.time())
	url_timestamp[parent_url] = x
	prev_url = parent_url
	#print("Total viewtimes: ", website_dict)

	with open('data.json', 'w') as file:
		json.dump(website_dict, file, indent=4)

	return jsonify({'message': parent_url + 'time logged'}), 200


@app.route('/scroll_update', methods=['POST'])
def scroll_update():
	resp_json = request.get_json()

	parent_url = url_domain(resp_json['url'])
	scroll_dist = resp_json['scroll_dist']

	print("Scroll recorded on: " + url_domain(parent_url) + ", dist = " + str(scroll_dist))

	return jsonify({'message': parent_url + 'scroll logged'}), 200


app.run(host='0.0.0.0', port=5000)

extension_website_tracker/content.js
var max_y = 0

function updateScroll() {
    var doc = document, w = window;
    var x, y, docEl;
    
    if ( typeof w.pageYOffset === 'number' ) {
        x = w.pageXOffset;
		y = w.pageYOffset;
    } else {
        docEl = (doc.compatMode && doc.compatMode === 'CSS1Compat')? doc.documentElement: doc.body;
        x = docEl.scrollLeft;
		y = docEl.scrollTop;
    }
	
	if(y > max_y)
		max_y = y;
}

window.addEventListener("scroll", updateScroll);

window.addEventListener("blur", function () {
	var xhttp = new XMLHttpRequest();
	xhttp.onreadystatechange = function () {
		if (this.readyState == 4 && this.status == 200) {
			console.log(this.responseText);
		}
	};
	xhttp.open("POST", "http://127.0.0.1:5000/scroll_update");
	xhttp.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
	xhttp.send(JSON.stringify({"url": this.window.location.href, "scroll_dist": max_y}));
	//xhttp.send(JSON.stringify({"scroll_dist": y_scroll}));
	//xhttp.send("hello")
})

extension_website_tracker/.ipynb_checkpoints/background-checkpoint.js
chrome.tabs.onCreated.addListener((tab) => {
	chrome.tabs.executeScript(tab.id, {file: "jquery.js"}, function() {
		chrome.tabs.executeScript(tab.id, {file: "content.js"});
	});
})

chrome.tabs.onActivated.addListener(function (activeInfo) {
    chrome.tabs.get(activeInfo.tabId, function (tab) {
        y = tab.url;
        var xhttp = new XMLHttpRequest();
        xhttp.onreadystatechange = function () {
            if (this.readyState == 4 && this.status == 200) {
                console.log(this.responseText);
            }
        };
        xhttp.open("POST", "http://127.0.0.1:5000/send_url");
		xhttp.send("url_for_time=" + y);
    });
});

chrome.tabs.onUpdated.addListener((tabId, change, tab) => {
    if (tab.active && change.url) {

        var xhttp = new XMLHttpRequest();
        xhttp.onreadystatechange = function () {
            if (this.readyState == 4 && this.status == 200) {
                console.log(this.responseText);
            }
        };
        xhttp.open("POST", "http://127.0.0.1:5000/send_url");
		xhttp.send("url_for_time=" + change.url);

    }
});

// chrome.tabs.executeScript({
// 	tabId: tabId,
// 	file: "content.js",
// 	code: code
// });

// // define a mapping between tabId and url:
// var tabToUrl = {};

// chrome.tabs.onUpdated.addListener(function (tabId, changeInfo, tab) {
// 	//store tabId and tab url as key value pair:
// 	tabToUrl[tabId] = tab.url;
// });

// chrome.tabs.onRemoved.addListener(function (tabId, removeInfo) {
// 	//since tab is not available inside onRemoved,
// 	//we have to use the mapping we created above to get the removed tab url:
// 	console.log(tabToUrl[tabId]);

// 	var xhttp2 = new XMLHttpRequest();
// 	xhttp2.onreadystatechange = function () {
// 		if (this.readyState == 4 && this.status == 200) {
// 			console.log(this.responseText);
// 		}
// 	};
// 	xhttp2.open("POST", "http://127.0.0.1:5000/quit_url");
// 	xhttp2.send("url=" + tabToUrl[tabId]);

// 	// Remove information for non-existent tab
// 	delete tabToUrl[tabId];

// });

flask_server/.ipynb_checkpoints/server-checkpoint.py
from flask import Flask, jsonify,request
import time, tldextract, json, requests
import pickle
import pprint
from IPy import IP
from collections import defaultdict
import itertools  #used to slice a dictionary

app = Flask(__name__)

path_history = './../data/history.data'

try:
	with open(path_history, 'rb') as f:
		website_dict = pickle.load(f)
except:
	website_dict = {}

url_timestamp = {}
prev_url = ""

skip_domains = [
    'www.mail.google.com',
    'www.mail.yahoo.com',
    'www.account.google.com',
    'www.facebook.com',  
    'www.docs.google.com'
]

def isIP(str):
    try:
        IP(str)
    except ValueError:
        return False
    return True


def clean_url(url):
	url_list = tldextract.extract(url)
	domain_name = url_list.domain + '.' + url_list.suffix

	if(domain_name == '.'):
		return "EMPTY_DOMAIN_NAME"
	
	spltAr = url.split("://");
	protocol = spltAr[0]
	if protocol not in ['http', 'https']:
		return "DO_NOT_TRACK"
	i = (0,1)[len(spltAr)>1];
	dm = spltAr[i].split("?")[0].split('/')[0].split(':')[0].lower();
	if dm[0:4] != 'www.':
		# check if domain is ip address
		if isIP(dm):
			return "DO_NOT_TRACK"
		dm = 'www.' + dm
	# domain checks
	if dm in skip_domains:
		return "DO_NOT_TRACK"
	return url
	

dont_track = ["EMPTY_DOMAIN_NAME", "newtab.", "DO_NOT_TRACK"]


@app.route('/send_url', methods=['POST'])
def send_url():
	resp_json = request.get_data()
	params = resp_json.decode()
	url = params.replace("url_for_time=", "")
	print("Open website: " + clean_url(url))
	parent_url = clean_url(url)

	global url_timestamp
	global website_dict
	global prev_url

	if(parent_url not in website_dict.keys()) and (parent_url not in dont_track):
		website_dict[parent_url] = {'time_spent' : 0,
									   'scroll_dist' : 0}

	if(prev_url != '') and (prev_url not in dont_track):
		time_spent = int(time.time() - url_timestamp[prev_url])
		website_dict[prev_url]['time_spent'] += time_spent

	#print('Time on ' + prev_url + ' : ' + str(website_dict[prev_url]['time_spent']))

	x = int(time.time())
	url_timestamp[parent_url] = x
	prev_url = parent_url



	with open(path_history, 'wb') as f:
		 pickle.dump(website_dict, f)
			
	pprint.pprint(website_dict)

	return jsonify({'message': parent_url + 'time logged'}), 200


@app.route('/scroll_update', methods=['POST'])
def scroll_update():
	resp_json = request.get_json()

	parent_url = clean_url(resp_json['url'])
	scroll_dist = resp_json['scroll_dist']

	if(parent_url not in website_dict.keys()) and (parent_url not in dont_track):
		website_dict[parent_url] = {'time_spent' : 0,
									   'scroll_dist' : scroll_dist}
	else:
		website_dict[parent_url]['scroll_dist'] += scroll_dist

	print("Scroll on: " + clean_url(parent_url) + " : " + str(scroll_dist))

	return jsonify({'message': parent_url + 'scroll logged'}), 200


app.run(host='0.0.0.0', port=5000)